lib/go/thrift/response_helper.go

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package thrift
+
+import (
+	"context"
+)
+
+// See https://godoc.org/context#WithValue on why do we need the unexported typedefs.
+type responseHelperKeyType int
+
+// Value for responseHelperKey.
+const responseHelperKey = responseHelperKeyType(0)
+
+// TResponseHelper defines a object with a set of helper functions that can be
+// retrieved from the context object passed into server handler functions.
+//
+// Use GetResponseHelper to retrieve the injected TResponseHelper implementation
+// from the context object.
+type TResponseHelper interface {
+	// THeader related functions
+	THeaderResponseHelper
+}
+
+type responseHelper struct {
+	THeaderResponseHelper
+}
+
+type nopResponseHelper struct {
+	nopHeaderResponseHelper
+}
+
+var nopHelper TResponseHelper = nopResponseHelper{}
+
+// THeaderResponseHelper defines THeader related TResponseHelper functions.
+type THeaderResponseHelper interface {
```

Comment:
Not sure if we need the interface at all. What if we export `headerHelper` and let it handle the no-op case? For example:

```go
type HeaderHelper struct {
	proto *THeaderProtocol
}

func (h *HeaderHelper) SetHeader(key, value string) {
        if h != nil {
	    h.proto.SetWriteHeader(key, value)
        }
}

func (h *HeaderHelper) ClearHeaders() {
        if h != nil {
	    h.proto.ClearWriteHeaders()
        }
} 
```

Which would allow us to simplify the getter:
```go
func GetResponseHelper(ctx context.Context) (helper *HeaderHelper, ok bool) {
	if v := ctx.Value(responseHelperKey); v != nil {
		helper, ok = v.(TResponseHelper)
	}
	return
}
```

Reply:
We do need an interface if we want to embed it into the `TResponseHelper` interface, because you cannot embed a struct into an interface. I think we do want embedding because that way makes `TResponseHelper` extensible in the future.

But I like the idea of not having a separated `nopHeaderHelper`, and made that change. I'm not sure if we want to export `headerHelper` type (and `newHeaderHelp` function). Looking at the existing library code we didn't seem very consistent on whether to export those helper implementations. I didn't export them now but I don't feel strongly either way, but if we do decide to export them, we might need a better name (`HeaderHelper` sounds too generic, but we do need the name `THeaderResponseHelper` for the interface, not the concrete type).

Reply:
> We do need an interface if we want to embed it into the `TResponseHelper` interface

Only if we want `TResponseHelper` to be an interface. Why can't it be a struct? I think it's better if callers define interfaces if/when they need them, instead of the library exporting it.

In this case, we only provide optional helper functions so if the user wants a different implementation, they can write one and perhaps define their own interface which is implemented by our `TResponseHelper` as well.

> (`HeaderHelper` sounds too generic, but we do need the name `THeaderResponseHelper` for the interface, not the concrete type)

Not having interfaces would solve this as well.

Reply:
OK I converted interfaces into structs. PTAL.

Reply:
Looks good!